### PR TITLE
Feature/5 add extended transformations

### PR DESCRIPTION
**Pull request by yruslan**
To who would like to review this PR: new changes are in `Extensions.scala` and `README.md`. The rest is almost 100% from Enceladus.

### CODE REVIEW

File: src/main/scala/za/co/absa/spark/hats/Extensions.scala [hunk 1]
```diff
@@ -45,6 +47,19 @@ object Extensions {
...
 
+    /**
+      * Add a column that can be inside nested structs, arrays and its combinations.
+      *
+      * @param newColumnName A column name to be created
+      * @param expression    A new column value
+      * @return A dataframe with a new field that contains transformed values.
+      */
+    def nestedWithColumnExtended(newColumnName: String,
+                         expression: ExtendedTransformFunction): Dataset[Row] = {
```
Comment (Zejnilovic) [inline src/main/scala/za/co/absa/spark/hats/Extensions.scala]:
```suggestion
                                 expression: ExtendedTransformFunction): Dataset[Row] = {
```

**Review by Zejnilovic — COMMENTED** [review]
I went over it. I haven't got the skill and/or time yet to judge it fully. But I did find 1 spacing "issue" 😄

File: README.md [hunk 2]
```diff
@@ -83,8 +84,50 @@ scala> df.nestedWithColumn(newColumnName = "my_array.c", expression = lit("hello
 +---+---------------------------------------------------+
 ```
 
+### Add column (extended)
+The extended API method `nestedWithColumnExtended` works similar to the basic one, but allows the caller to reference
```
Comment (GeorgiChochov) [inline README.md]:
```suggestion
The extended API method `nestedWithColumnExtended` works similarly to the basic one but allows the caller to reference
```

File: README.md [hunk 3]
```diff
@@ -83,8 +84,50 @@ scala> df.nestedWithColumn(newColumnName = "my_array.c", expression = lit("hello
...
+other array elements, possibly on different levels of nesting. The way it allows this is a little tricky.
+The second parameter is changed from being a column to a *function that returns a column*. Moreover, this function has
+an argument which is a function itself, the `getField()` function. The `getField()` function can be used in the
+transformation to reference other columns in the dataframe by their fully qualified name.
+
+Int the following example, a transformation adds a new field `my_array.c` to the dataframe by concatenating a root
```
Comment (GeorgiChochov) [inline README.md]:
```suggestion
In the following example, a transformation adds a new field `my_array.c` to the dataframe by concatenating a root
```

File: README.md [hunk 4]
```diff
@@ -62,10 +59,14 @@ scala> df.show(false)
 ```
 
 ### Add a column
-The **nestedWithColumn** method allows adding new fields inside nested structures and arrays.
+The `nestedWithColumn` method allows adding new fields inside nested structures and arrays.
+
+The addition of a column API is provided in two flavors: the basic and the extended API. The basic API is simpler to
+use, but expressions it expects cannot reference columns on parent array levels. Here is an example of the basic add
```
Comment (GeorgiChochov) [inline README.md]:
```suggestion
use, but the expressions it expects can only reference columns inside the array and the root of the schema. Here is an example of the basic add
```

**Comment by yruslan** [comment]
Thanks for the review guys!

**Review by GeorgiChochov — APPROVED** [review]

**Pull request merged**